### ISSUE TITLE
Godot 3.2 Alpha 3 C# Issue when exporting a C# project to Android with 4x MSAA forced on in your smartphone developer settings

### ISSUE THREAD

**Issue by TheLuigiplayer** [godotengine/godot#33223]
**Godot version:**
v3.2.alpha3.mono.official

**OS/device including version:**
Windows 10 Version 1903

**Issue description:**
When using your Android Smartphone as the Test Device for a C# project, the app will boot but crash during the Godot Splash Screen. It will show the Splash Screen and then the Splash Screen will start to flicker and nothing else happens. However, this seems to only be an issue, when you force 4xMSAA to OpenGLES 2.0 apps in your smartphone and export your Godot App for GLES3.0. Here are my tests using a OnePlus 6T:
    1.OP6T 4x MSAA disabled + Godot GLES3 + Godot 4x MSAA disabled = Succeed
    2.OP6T 4x MSAA disabled + Godot GLES3 + Godot 4x MSAA enabled = Succeed
    3.OP6T 4x MSAA disabled + Godot GLES2 + Godot 4x MSAA disabled= Succeed
    4.OP6T 4x MSAA disabled + Godot GLES2 + Godot 4x MSAA enabled= Succeed
    5.OP6T 4x MSAA enabled + Godot GLES3 + Godot 4x MSAA disabled = Failure
    6.OP6T 4x MSAA enabled + Godot GLES3 + Godot 4x MSAA enabled = Failure
    7.OP6T 4x MSAA enabled + Godot GLES2 + Godot 4x MSAA disabled = Succeed
    8.OP6T 4x MSAA enabled + Godot GLES2 + Godot 4x MSAA enabled = Succeed
With GLES2 Fallback:
    9. OP6T 4x MSAA enabled + Godot GLES3 + Godot GLES2 Fallback + Godot 4x MSAA disabled = Failure
    10. OP6T 4x MSAA enabled + Godot GLES3 + Godot GLES2 Fallback + Godot 4x MSAA enabled = Failure

**Steps to reproduce:**
-Start a new C# project
-Add atleast one C# script
-Configure your Godot Editor and your Android Smartphone, that you can use it as the Test Device for the Editor
-Force 4xMSAA on, in your smartphone developer settings
-Export your Godot App for GLES3.0
-Test your Scene on your Android Smartphone

**Minimal reproduction project:**
-Test your GLES3.0 C# project on your Android Smartphone with 4xMSAA forced on.

**Comment by volzhs** [comment]
you can find what was wrong with `adb logcat`

**Comment by FeralBytes** [comment]
@TheLuigiplayer A flickering screen may mean that your Android device does not support all of GLES3 that Godot is using, can you make sure that you are using Godot in GLES2 mode with ETC compression enabled on export?

**Comment by TheLuigiplayer** [comment]
> 
> 
> @TheLuigiplayer A flickering screen may mean that your Android device does not support all of GLES3 that Godot is using, can you make sure that you are using Godot in GLES2 mode with ETC compression enabled on export?

I kinda found a solution after some hours of testing. Seems like there was a problem with my OnePlus 6t when forcing 4x MSAA for GLES2 applications, although I'm exporting for GLES3. I turned it off and it seems to work now. I've also tested when forcing 4x MSAA on my OnePlus 6T, setting the project's driver to GLSE2 and activating 4x MSAA in the project itself, this worked aswell. So there seems to be a problem with MSAA and GLSE2/GLSE3 compatibility.

**Comment by FeralBytes** [comment]
Ok, can you break out a chart of what combinations resulted in the failing states, and which combinations worked? That will make it clearer. Additionally for both the failing and succeeding states can you do an adb logcat capture? I know that is a lot of work, but it helps with debugging, since most of us probably do not own a [OnePlus 6T](https://www.oneplus.com/6t/specs).

**Comment by TheLuigiplayer** [comment]
> 
> 
> Ok, can you break out a chart of what combinations resulted in the failing states, and which combinations worked? That will make it clearer. Additionally for both the failing and succeeding states can you do an adb logcat capture? I know that is a lot of work, but it helps with debugging, since most of us probably do not own a [OnePlus 6T](https://www.oneplus.com/6t/specs).

Sure!

1. OP6T 4x MSAA disabled + Godot GLES3 + Godot 4x MSAA disabled = Succeed
2. OP6T 4x MSAA disabled + Godot GLES3 + Godot 4x MSAA enabled = Succeed
3. OP6T 4x MSAA disabled + Godot GLES2 + Godot 4x MSAA disabled= Succeed
3. OP6T 4x MSAA disabled + Godot GLES2 + Godot 4x MSAA enabled= Succeed
4. OP6T 4x MSAA enabled + Godot GLES3 + Godot 4x MSAA disabled = Failure
5. OP6T 4x MSAA enabled + Godot GLES3 + Godot 4x MSAA enabled = Failure
6. OP6T 4x MSAA enabled + Godot GLES2 + Godot 4x MSAA disabled = Succeed
7. OP6T 4x MSAA enabled + Godot GLES2 + Godot 4x MSAA enabled = Succeed

(With GLES2 Fallback)
8. OP6T 4x MSAA enabled + Godot GLES3 + Godot GLES2 Fallback + Godot 4x MSAA disabled = Failure
9. OP6T 4x MSAA enabled + Godot GLES3 + Godot GLES2 Fallback + Godot 4x MSAA enabled = Failure

The adb logcat I've found doesn't seem to work, as it lists all my mobile acitivites. is there a way, that it will only show what happens with Godot Apps?

**Comment by FeralBytes** [comment]
```shell
adb logcat -s godot -s DEBUG -s GODOT
```
But beware, a low level failure, like this, may require more than just what Godot spits out. As it may be important to see what the phone spits out for a backtraces, if the java activity fails as a whole. SO the best way may be to have you manually filter out the other activities. Something else that can help is to throw the phone into developer mode and then choosing the developer option of "Limit background processes" to "no background processes", just while running adb. But don't forget to set it back to standard limit when you are done.
Another note my phone has a note that says it's "4x MSAA" option is only for Open GLES 2.0 applications, but my phone is older than yours, so it may mean nothing.

**Comment by TheLuigiplayer** [comment]
Yes I've looked into it again and it says, that it forces "4x MSAA" when running GLES 2.0 apps. The ADB Log really doesn't say much about the failure itself. Still I've attached the log to this reply.
[failure.log](https://github.com/godotengine/godot/files/3800068/failure.log)

**Comment by FeralBytes** [comment]
@TheLuigiplayer  can you upload this "Logfile is: /data/data/com.theluigiplayer.testgame/files/mono/mono_logs/2019_11_02 03.29.49 (24384).txt"  crash log. It has a weird message about API hash mismatch, and that may get us closer to the exact cause.

**Comment by TheLuigiplayer** [comment]
@FeralBytes Somehow, eventhough it says that a logfile is created I can't find it. It's not on my smartphone and neither on my PC. Also, there isn't even a package created on my smartphone.

**Comment by FeralBytes** [comment]
@TheLuigiplayer /data/data/ is definitely on the phone. It will be ok in the package/ APK install area for your game. I think unfortunately to get at it you might have to be root, thus a rooted phone.

**Comment by TheLuigiplayer** [comment]
@FeralBytes Okay I was able to acces it with Android Studio.
[2019_11_06 19.18.12 (4061).txt](https://github.com/godotengine/godot/files/3816107/2019_11_06.19.18.12.4061.txt)

**Comment by FeralBytes** [comment]
Below seems to be the main issue. But not sure how to fix it, hoping one of the other devs will take a look.
```
AOT: image '.mono/assemblies/Debug/mscorlib.dll.so' not found: dlopen failed: library ".mono/assemblies/Debug/mscorlib.dll.so" not found (in domain Mono, info)
AOT: image '/data/app/com.lamazstudios.testgame-YYrplXxWXKYDHwPJGIpZ-Q==/lib/arm64/mono/aot-cache/arm64/mscorlib.dll.so' not found: dlopen failed: library "/data/app/com.lamazstudios.testgame-YYrplXxWXKYDHwPJGIpZ-Q==/lib/arm64/mono/aot-cache/arm64/mscorlib.dll.so" not found (in domain Mono, info)
```
I am not normally a C# coder, I just started expirmenting with it. But something tells me that mscorlib is very important to C#.

**Comment by Calinou** [comment]
Does this occur in a project that only uses GDScript (using non-Mono-enabled export templates)?

**Comment by KoBeWi** [comment]
Is this issue still valid in Godot 4.4?

**Comment by FeralBytes** [comment]
@KoBeWi Given how long ago this was in mobile device terms,that there were lower end devices during that time, and that Godot 4.4 is targeted towards higher end devices. I do not feel this is still relevant. Unless @TheLuigiplayer joins back in the discussion with something I am missing it can be closed so are as I am concerned. This was some rather edge case stuff at the time.

**Comment by KoBeWi** [comment]
Closing then, comment if it's still relevant.
